File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.js
function solution(bridge_length, weight, truck_weights) {
    let second = 0;
    let next_truct = 0;
    let current_weight = 0;
    let bridge = new Array(bridge_length).fill(0);
    
    while(truck_weights.length !== 0) {
        second++;
        
        // truct escape bridge
        if(bridge[0] != 0) {
            current_weight -= bridge[0];
        }
        
        // truck move forward
        bridge.shift();
        bridge.push(0);
        
        next_truct = truck_weights[0];
        // next truck on bridge
        if(next_truct + current_weight <= weight) {
            bridge[bridge.length - 1] = truck_weights.shift();
            current_weight += next_truct;
        }
    }
    return second + bridge_length;
}

File: 프로그래머스/2/42583. 다리를 지나는 트럭/다리를 지나는 트럭.py
from collections import deque

def solution(bridge_length, weight, truck_weights):
    time = 0
    total_weight = 0
    bridge = deque([0] * bridge_length)
    
    for truck in truck_weights:
        while True:
            time += 1
            total_weight -= bridge.popleft()
            
            if total_weight + truck <= weight:
                bridge.append(truck)
                total_weight += truck
                break
            else:
                bridge.append(0)
                
    return time + bridge_length
    